pkg/operation/botanist/nodelocaldns.go

```diff
@@ -54,27 +53,24 @@ func (b *Botanist) DefaultNodeLocalDNS() (nodelocaldns.Interface, error) {
 		nodeLocalDNSForceTcpToUpstreamDNS = forceTcp
 	}
 
-	values := nodelocaldns.Values{
-		Image:                 image.String(),
-		VPAEnabled:            b.Shoot.WantsVerticalPodAutoscaler,
-		ForceTcpToClusterDNS:  nodeLocalDNSForceTcpToClusterDNS,
-		ForceTcpToUpstreamDNS: nodeLocalDNSForceTcpToUpstreamDNS,
-		ClusterDNS:            clusterDNS,
-		DNSServer:             dnsServer,
-	}
 	return nodelocaldns.New(
 		b.K8sSeedClient.Client(),
 		b.Shoot.SeedNamespace,
-		values,
+		nodelocaldns.Values{
+			Image:                 image.String(),
+			VPAEnabled:            b.Shoot.WantsVerticalPodAutoscaler,
+			ForceTcpToClusterDNS:  nodeLocalDNSForceTcpToClusterDNS,
+			ForceTcpToUpstreamDNS: nodeLocalDNSForceTcpToUpstreamDNS,
+			ClusterDNS:            clusterDNS,
+			DNSServer:             dnsServer,
+		},
 	), nil
 }
 
-// DeployNodeLocalDNS deploys the node-local-dns system component.
-func (b *Botanist) DeployNodeLocalDNS(ctx context.Context) error {
-	return b.Shoot.Components.SystemComponents.NodeLocalDNS.Deploy(ctx)
-}
-
-// DestroyNodeLocalDNS deletes the managed resource node-local-dns.
-func (b *Botanist) DestroyNodeLocalDNS(ctx context.Context) error {
-	return client.IgnoreNotFound(managedresources.Delete(ctx, b.K8sSeedClient.Client(), b.Shoot.SeedNamespace, "shoot-core-node-local-dns", false))
+// ReconcileNodeLocalDNS deploy/destroy node-local-dns component.
+func (b *Botanist) ReconcileNodeLocalDNS(ctx context.Context) error {
+	if b.Shoot.NodeLocalDNSEnabled {
+		return b.Shoot.Components.SystemComponents.NodeLocalDNS.Deploy(ctx)
+	}
+	return client.IgnoreNotFound(managedresources.Delete(ctx, b.K8sSeedClient.Client(), b.Shoot.SeedNamespace, nodelocaldns.ManagedResourceName, true))
```

Comment:
Again, why don't you call the `Destroy` function of the `NodeLocalDNS` struct here?

Reply:
We can't use, basically every time shoot reconcile this Destroy will be called if `node-local-dns` is not active and in that case, NodeLocalDNS interface is set to null and we can't call any function of that interface as it will create panic.

Reply:
Then just call create the `NodeLocalDNS` structure independent of whether the shoot enables it. This is better than duplicating implementation details of the `nodelocaldns` package here. WDYT?

Reply:
I would suggest using `client.IgnoreNotFound(managedresources.Delete(ctx, b.K8sSeedClient.Client(), b.Shoot.SeedNamespace, nodelocaldns.ManagedResourceName, true))` in the Destory function of `NodeLocalDNS` interface instead of `managedresources.DeleteForShoot(ctx, c.client, c.namespace, ManagedResourceName)`.

Reply:
I'm confused, how is this related tot his discussion? Anyways, why are you suggesting this now/what is the difference?

Reply:
I will go with the first idea `create the NodeLocalDNS structure independent of whether the shoot enables it.` cause this also is dplicating.

> I would suggest using `client.IgnoreNotFound(managedresources.Delete(ctx, b.K8sSeedClient.Client(), b.Shoot.SeedNamespace, nodelocaldns.ManagedResourceName, true))` in the Destory function of `NodeLocalDNS` interface instead of `managedresources.DeleteForShoot(ctx, c.client, c.namespace, ManagedResourceName)`.